### ISSUE TITLE
Feature Request: Restrict BayLeaf API to open-weight models once OpenRouter supports it as a routing toggle

### ISSUE THREAD

**Issue by rndmcnlly** [rndmcnlly/bayleaf#25]
## Context

BayLeaf already enforces zero-data-retention (ZDR) routing for all inference via OpenRouter. This is stated prominently in the README and POSITION.md, and it is a baseline criterion the project holds other tools to as well.

The current ZDR enforcement works because OpenRouter offers a simple account-level toggle at Settings → Privacy: flip it on, and your API key can only route to ZDR endpoints. No manual allowlist maintenance, no risk of accidental misrouting.

## The gap

BayLeaf's "Basic" model is explicitly described as "a large open-weight LLM" (chat/models/basic/model.json). The POSITION.md argues that model-provider switchability and open-weight inference are architectural choices, not fantasies. But right now, there is no OpenRouter-level toggle to restrict routing to open-weight models only — the way there is for ZDR.

If someone (a student, a script, an integration) uses a BayLeaf API key with an explicit closed-weight model slug (e.g. `openai/gpt-4`), the request would route through — because nothing at the OpenRouter level prevents it. The BayLeaf proxy could of course validate model slugs client-side, but that's a manually maintained allowlist, brittle in the same way a ZDR allowlist would be brittle without the toggle.

## The ask

**Once OpenRouter ships an account-level "open-weight only" toggle** (or equivalent Guardrail/parameter — see [OpenRouterTeam/docs#2](https://github.com/OpenRouterTeam/docs/issues/2)), flip it on for the BayLeaf OpenRouter account, and note the restriction in the README alongside the existing ZDR claim.

The README line:

> All LLM inference routes through **zero-data-retention (ZDR)** providers via OpenRouter.

would become something like:

> All LLM inference routes through **zero-data-retention (ZDR)** providers and **open-weight models** via OpenRouter.

The POSITION.md baseline criteria could also be updated to make open-weight inference an explicit fifth criterion, bringing the practice in line with the philosophy that is already articulated but not yet enforced at the routing layer.

## Why wait for OpenRouter

I filed [OpenRouterTeam/docs#2](https://github.com/OpenRouterTeam/docs/issues/2) requesting exactly this feature — an account-level toggle (and per-request parameter) that restricts routing to open-weight models, mirroring the existing ZDR toggle pattern. The argument is straightforward: the metadata already exists in OpenRouter's model catalog, the enforcement architecture is proven (ZDR uses it), and a property-based toggle eliminates the brittleness of manual allowlists.

Until that toggle exists, a client-side model-slug allowlist in the BayLeaf proxy would be the only option — but that's a maintenance burden that grows with every model addition and removal, and it doesn't align with the "flip a switch, it just works" experience BayLeaf already has for ZDR. Better to wait for the proper mechanism than to build a temporary one that will be abandoned.

## Dependency

- [OpenRouterTeam/docs#2](https://github.com/OpenRouterTeam/docs/issues/2) — Feature Request: Account-level toggle to restrict routing to open-weight models

**Comment by rndmcnlly** [comment]
Closing because I don't think OR is ever going on act on this, and I don't want to clutter my issues list.

**Comment by rndmcnlly** [comment]
Reopening: the premise this was closed on turns out to be wrong, and in our favor.

I closed this in May with "I don't think OR is ever going to act on this." That's still true — [OpenRouterTeam/docs#2](https://github.com/OpenRouterTeam/docs/issues/2) is **still open, 3+ months later, with no response**, no label, no rejection. So the *dependency* hasn't moved.

What has changed is that the dependency turns out to be unnecessary. The original issue said "the metadata already exists in OpenRouter's model catalog" as an argument for why the toggle *should* be easy for them to build. That was right, and I under-read my own claim: the metadata doesn't just exist internally, **it's already exposed on the public `/api/v1/models` response.** Every model object carries a `hugging_face_id` field. That's a usable open-weight predicate today, and it's much closer to "flip a switch" than to the brittle hand-maintained allowlist this issue was closed to avoid.

## What I verified against the live catalog

Pulled `https://openrouter.ai/api/v1/models` (367 models). `hugging_face_id` is populated for **152/367 (41%)**.

First, confirming the toggle really doesn't exist, so nobody has to re-check:

- Documented query params on `/api/v1/models` are only `output_modalities`, `supported_parameters`, `sort`, `offset`/`limit`. No open-weight filter.
- `?open_weights=true` and `?license=open` both return all 367 — **unknown params are silently ignored**, not rejected. Worth knowing: a filter that looks like it works and doesn't. (`?category=open` does error, with a ZodError listing only content categories like `programming`/`roleplay`.)
- No `/collections/open-weights` — 404s identically to a nonsense slug I tested as a control.
- Guardrails do offer `allowed_models` / `blocked_models` / `allowed_providers`, but those are enumerations you maintain. Same burden as doing it ourselves, just stored in someone else's dashboard.

Then, grading the predicate against ground truth:

| Test | Result |
|---|---|
| Current recommended + Pareto: `z-ai/glm-5.2`, `deepseek/deepseek-v4-pro`, `deepseek-v4-flash`, `moonshotai/kimi-k3` | ✅ all populated |
| Provenance-diverse candidates: `nvidia/nemotron-3-ultra-550b-a55b`, `openai/gpt-oss-120b`, `google/gemma-4-*`, `thinkingmachines/inkling`, `minimax/minimax-m3` | ✅ all populated |
| `anthropic/*`, `openai/gpt-*`, `google/gemini-*`, `x-ai/*` | ✅ correctly null |
| False negatives (open-weight but null) | **2** — `mistralai/mistral-large-2512` (Apache 2.0, 675B/41B active), `minimax/minimax-m1` |

Two things make this better than a vendor allowlist, and they're the reason I think this clears the bar the issue set:

**It keys on the model, not the vendor.** `openai/gpt-oss-120b` and `gpt-oss-20b` are populated — open-weight models from an otherwise closed-weight vendor. A vendor-prefix allowlist gets that wrong.

**It correctly excludes the API-only siblings of open families.** This is the part that would have bitten us. `z-ai/glm-5-turbo` and `glm-5v-turbo` are null while `glm-5.2` / `glm-5.1` / `glm-5` are populated. Same split for `qwen3.7-max`, `qwen3-max`, `qwen3.6-plus`, `qwen3.6-flash`, and `mistral-large-2407`. And `qwen/qwen3-coder-plus` self-describes as "Alibaba's **proprietary** version of the Open Source Qwen3 Coder 480B A35B" — correctly excluded. A `qwen/*` + `mistralai/*` + `z-ai/*` allowlist of the kind this issue was closed to avoid would have quietly served proprietary models under an open-weight policy. That's the failure mode that turns a stated commitment into an embarrassment, and this field encodes the distinction more accurately than I would by hand.

## Proposed implementation

Filter on `hugging_face_id != null` in the existing `/v1/models` interceptor in `api/src/routes/proxy.ts` — right where we already `.map()` to add the `openrouter:` prefix. Optionally a 2-entry additive override for the false negatives above. That's a ~2-line exception list against a live-updating 152-model base, not a 150-entry allowlist re-audited monthly. It composes with the `bedrock:` mantle path, which is already live-fetched and open-weight-only.

Three caveats that belong in the code comments and the public docs:

1. **It's a weights-availability proxy, not a license assertion.** It says a HF repo exists. It does not say the license is OSI-approved, and says nothing about training data — the open-weight vs. open-source line that OSI is (rightly) pushing on. `politics/README.md` commitment #2 already links HuggingFace and says "available for anyone to download, audit, or build on," so the field measures approximately the thing we claimed. Public wording should be "models OpenRouter reports as having published weights on HuggingFace," not "we audited the licenses."
2. **It's undocumented.** OpenRouter could stop populating it without notice, and per the silent-param-ignoring behavior above, we'd get no error — the list would just quietly stop matching our stated policy.
3. **Fail closed.** If the predicate degrades to null-for-everything we'd list zero models. Given essentially all API traffic is on the recommended default and routing is untouched, an empty list is annoying and harmless, and it's the correct failure direction for a values commitment. Matches the existing `ALT_BACKENDS` fail-closed posture.

## Scope note: listing ≠ routing

The original ask was routing enforcement ("a request with `openai/gpt-4` would route through"). This gets us **listing** enforcement, which is a narrower thing, and I now think listing-only is the right first step rather than a compromise:

- Essentially all API traffic already uses the recommended `openrouter:z-ai/glm-5.2` default, so **nobody gets cut off**. The change is expressive, not operational.
- It resolves a live contradiction in this repo. `README.md` currently says the API "allow[s] optional access to proprietary models for tasks that warrant them" while `politics/README.md` commitment #2 says BayLeaf "**exclusively** serves open-weight models." Both public, mutually inconsistent.
- Keeping bare-slug passthrough working preserves comparative model research — a student doing an AI-evaluation project, or a colleague studying open-vs-closed behavioral differences, needs to reach proprietary models. Breaking that in a project that calls itself *counterfoil research* would be self-undermining. Curation expresses the politics; the escape hatch protects inquiry. That carve-out should be documented as deliberate, not left looking like an implementation gap.

## Two things to fix in the same pass

Both surfaced while working this through, and both would otherwise become fresh contradictions created *by* the fix:

**Commitment #1 vs. the open frontier.** `politics/README.md` commits to "mid-sized models (tens to hundreds of billions of parameters), not the trillion-parameter flagships," and Basic's prompt says "sub-trillion-parameter." But the competitive open tier is now trillion-scale MoE — GLM-5.2 is ~744B total, and Kimi K3 / DeepSeek V4 Pro are covered as trillion-scale. The politics doc treats commitments #1 and #2 as mutually reinforcing; in 2026 they pull against each other. The honest repair is to restate #1 in terms of **active parameters per token** rather than total, which is the actual energy story for sparse MoE and a defensible claim. Leaving a total-parameter commitment in the manifesto while serving trillion-scale MoE is a grep away from being noticed.

**Provenance concentration.** The Pareto is Chinese labs at every tier (DeepSeek, Z.ai, Moonshot), and `OPENCODE_CURATED_MODELS` is currently `openrouter:qwen/qwen3.6-27b,vertex:zai-org/glm-5-maas` — where the Vertex entry is stripped anyway under `VERTEX_ENABLED: "false"`, leaving one Chinese model plus a Chinese recommended default. Codifying "open-weight only" today therefore means codifying "Chinese-lab only" in practice, as a **public university** service, while Texas has added DeepSeek to a state prohibited-technologies list under directives naming public institutions of higher education, a House probe into US corporate use of Chinese models is underway, and Axios reports the administration weighing Entity List additions and procurement rules. Note `politics/HECVAT.md` INTL-04 currently answers "Will data be collected from or processed in or stored in China?" with a flat "No" — still true (US providers, ZDR endpoints, no PRC processing), but the question a reviewer asks changes shape once provenance is written policy rather than incidental.

The fix is not to abandon the policy — it's provenance diversity *inside* it. Write the policy as **open-weight**, never as a named model set, and keep at least one non-Chinese open-weight model listed per tier. `nvidia/nemotron-3-ultra-550b-a55b` (#2 on Artificial Analysis's open-weight index, ~3 points behind GLM-5.2) and `openai/gpt-oss-120b` are both live and both pass the predicate, so seeding that is a one-line env change. Then a procurement rule landing means we **swap within policy, not out of it** — which is a live demonstration of `POSITION.md` criterion #4 (switchability) under a geopolitical trigger rather than a commercial one, and a better test of it than the OpenRouter/NRP story we currently cite.

## Framing, for whenever this gets written up

Two notes for the eventual README / POSITION.md edit:

- Prefer a **time-lag** claim over a quality-tier claim: OpenRouter's own June 2026 analysis puts open weights at a "consistent 3-6 month gap for over 18 months," and UK AISI independently measures GLM-5.2 at 4-7 months behind the closed frontier on cyber. "BayLeaf serves the frontier of about six months ago, deliberately, at roughly a tenth of the cost" survives leaderboard churn; "the open tier is competitive right now" expires on the next release. The Pareto/capability argument should be a footnote at most — it isn't load-bearing and including it invites relitigation every cycle.
- For the institutional audience, the July 24 2026 letter "Open Weights and American AI Leadership" (~50 signatories incl. Nvidia, Microsoft, Meta, Google, OpenAI) argues nearly verbatim what this issue wants: open weights let institutions "match the right model to the right job at the right cost, **reserving frontier-scale capability for genuine frontier problems**." Useful to be able to say the posture is what the largest vendors are lobbying Washington for. Use with care — the same letter is a defense against the restrictions described two sections up.

Not doing any of the above yet — reopening so the finding is on the record and the upstream dependency stops being a blocker.

---

*Findings verified against the live OpenRouter catalog on 2026-07-27 (367 models). Investigated with agent assistance via the Lathe toolkit, consistent with the repo's [GenAI disclosure](https://github.com/bayleaf-ucsc/bayleaf#genai-disclosure); the counts and per-model checks above are reproducible from `GET https://openrouter.ai/api/v1/models`.*

**Comment by rndmcnlly** [comment]
**Correction to the comment above — the proposed predicate is wrong, and it fails open.**

I wrote "filter on `hugging_face_id != null`." Do not implement that. While preparing a note for [OpenRouterTeam/docs#2](https://github.com/OpenRouterTeam/docs/issues/2) I checked the field's actual value distribution and found it has **three** states, not two:

| Value | Count (of 367) |
|---|---|
| non-empty string | **152** |
| empty string `""` | **127** |
| `null` | **88** |

Both `""` and `null` mean "no weights published" — the distinction between them appears to be incidental (which upstream path populated the record), not semantic. So:

- `hugging_face_id != null` → admits **279** models, wrongly including 127 with `""` — among them `openai/gpt-5.5`, `openai/gpt-5.5-pro`, `x-ai/grok-4.20`, and `z-ai/glm-5v-turbo`. **81 are unambiguously closed-weight.**
- `Boolean(hugging_face_id)` (truthy) → **152** models, matching the verified figure in the comment above.

The correct predicate is the **truthy** check:

```ts
.filter(m => Boolean(m.hugging_face_id))   // NOT != null
```

Two things worth drawing out, because they're the actual lesson rather than a typo:

**This is precisely the fail-open failure I described two sections up and then walked into anyway.** The comment above argues the danger of a vendor allowlist is that it "would quietly serve proprietary models under an open-weight policy." A `!= null` filter does exactly that, and does it *silently* — the endpoint returns 200, the list looks plausible, and `gpt-5.5` sits in it. My spot-checks passed because I verified that closed models were *falsy*, and `""` is falsy in JS, so the behavior I sampled was correct while the predicate I wrote was not.

**It strengthens caveat 2, rather than undermining the approach.** The field is in OpenRouter's OpenAPI spec (`openapi.yaml`, `hugging_face_id: type: [string, null]`, "Hugging Face model identifier, if applicable") — so it's more documented than I credited. But the spec declares only `string | null` and says nothing about `""`, which is in fact the *most common* falsy value. An implementation written to the published schema gets this wrong. That makes the canary in caveat 2 load-bearing rather than defensive, and it needs to assert falsiness specifically:

```ts
// must catch "" and null, not just null
assert(Boolean(model('z-ai/glm-5.2').hugging_face_id) === true);
assert(Boolean(model('openai/gpt-5.5').hugging_face_id) === false);
assert(Boolean(model('anthropic/claude-opus-5').hugging_face_id) === false);
```

Everything else in the comment above stands — the 152/367 count, the two false negatives, the model-not-vendor and turbo-vs-open findings, and the scope reasoning were all computed with a truthy check and are unaffected. Only the one-line `!= null` recommendation was wrong. Filing the `""`-vs-`null` inconsistency upstream on docs#2, since it's a concrete spec-vs-behavior mismatch independent of the feature request.

**Comment by rndmcnlly** [comment]
Implemented and closing. The predicate from the thread above (truthy `hugging_face_id`, not `!= null`) is live in prod as of 2026-07-29.

**What shipped:**

- `/v1/models` now lists only the 152 open-weight OpenRouter entries. Verified against the live catalog: gpt-5.5, claude-opus-5, grok-4.20, glm-5v-turbo all de-listed; gpt-oss-120b and nemotron-3-ultra remain listed (the model-not-vendor property holds). No override list: the two known false negatives (mistral-large-2512, minimax-m1) fail closed, OpenRouter's to fix upstream.
- Closed-weight models remain reachable by crafting the slug: the escape hatch scoped above, now documented as deliberate in llms.txt. Listing is curated, routing is open.
- **Plot twist: bedrock.** Re-auditing the bedrock slice against the same standard overturned #41's core assumption. Mantle's 47-entry catalog includes 7 closed-weight models (claude-haiku-4-5, gpt-5.4/5.6, grok-4.3, palmyra-vision) and exposes no weights-availability field to filter on. With no mechanical predicate and the backend getting very little use, we turned it off (`BEDROCK_ENABLED: "false"`).
- **ZDR enforcement hole found and closed**, in preparation for someday turning bedrock back on. Mantle's per-model `data_retention.allowed_modes` shows the gpt-5.x entries can never run zero-retention (no `'none'` mode). We set the AWS account's mantle mode to `'none'`, so non-ZDR models report `status: "unavailable"` and are blocked server-side, crafted slugs included. Verified live: gpt-5.4 invocation rejected (HTTP 400), gpt-oss-120b still serves (HTTP 200). Undocumented prerequisite: the API-key IAM user needs `bedrock-mantle:PutAccountDataRetention` granted explicitly (ships without it, 403 until added).
- Code quality in passing: `fetchBedrockModels` now filters the listing to ZDR-capable models and fails closed on missing metadata; stale backend advertising corrected in llms.txt and the OpenAPI route description.

The bedrock re-enable checklist (enterprise BAA, re-apply mode `'none'` + SCP, weights-listing answer, privacy.html subprocessor) lives in `api/wrangler.jsonc` and `api/AGENTS.md`. The canary from caveat 2 was consciously skipped: the fail-closed direction makes a degraded predicate an annoyance, not a leak.
